File: ts-service/src/subscribe.ts
import { connect } from './connection.js';
import { Validator } from './validate.js';
import { Subscriber } from './subscriber.js';
import { SPEC_PATH } from './spec-path.js';

const PATTERN = process.env.SUBSCRIBE_PATTERN ?? 'events.#';

async function main(): Promise<void> {
  const validator = new Validator(SPEC_PATH);
  const conn = await connect();
  const subscriber = new Subscriber(conn, validator, PATTERN);

  const shutdown = async () => {
    await conn.close().catch(() => {});
    process.exit(0);
  };
  process.on('SIGINT', shutdown);
  process.on('SIGTERM', shutdown);

  await subscriber.run();
}

main().catch((err) => {
  console.error('[ts-service] fatal:', err);
  process.exit(1);
});

File: ruby-service/subscribe.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Long-running subscriber. Binds to the topic exchange with SUBSCRIBE_PATTERN
# and validates every event it receives against the AsyncAPI contract.

require_relative 'boot'

SPEC_PATH = ENV.fetch('SPEC_PATH', File.expand_path('../asyncapi.yaml', __dir__))
PATTERN = ENV.fetch('SUBSCRIBE_PATTERN', 'events.#')

$stdout.sync = true

validator = Validator.new(SPEC_PATH)
connection = Connection.new
subscriber = Subscriber.new(connection, validator, pattern: PATTERN)

trap('INT')  { connection.close; exit }
trap('TERM') { connection.close; exit }

subscriber.run
